### ISSUE TITLE
[Inductor][CPU] torch.compile returns 0 for short positive uint8/int8 amin reductions on Apple Silicon

### ISSUE THREAD

**Issue by lackjava-creator** [pytorch/pytorch#191499]
### 🐛 Describe the bug

On an Apple Silicon CPU, Inductor returns `0` for a short positive `uint8` or `int8` `amin` reduction, while eager execution returns the correct minimum. This is a silent wrong-result issue and reproduces on the first compiled call with a fresh Inductor cache.

The problem occurs with both static and dynamic compilation. A four-element tensor is sufficient; lengths 1–3 pass in my local checks, while lengths 4 and above fail.

### Minimal reproducer

```python
import torch

print("torch:", torch.__version__)
print("git:", torch.version.git_version)

for dtype in (torch.uint8, torch.int8):
    x = torch.tensor([5, 6, 7, 8], dtype=dtype)

    for dynamic in (False, True):
        torch._dynamo.reset()
        compiled_fn = torch.compile(
            lambda value: torch.amin(value),
            backend="inductor",
            fullgraph=True,
            dynamic=dynamic,
        )

        eager = torch.amin(x)
        compiled = compiled_fn(x)
        print(dtype, "dynamic=", dynamic,
              "eager=", eager.item(),
              "compiled=", compiled.item())

        torch.testing.assert_close(compiled, eager, rtol=0, atol=0)
```

Observed on PyTorch 2.13.0:

```text
torch.uint8 dynamic=False eager=5 compiled=0
torch.uint8 dynamic=True  eager=5 compiled=0
torch.int8  dynamic=False eager=5 compiled=0
torch.int8  dynamic=True  eager=5 compiled=0
```

Expected: all compiled results are `5`.

### Isolation results

- Reproduced 3/3 with a new `TORCHINDUCTOR_CACHE_DIR` each time.
- `uint8` and `int8` fail; `int16`, `int32`, `int64`, `float16`, and `float32` pass for the same values.
- Setting `torch._inductor.config.cpp.simdlen = 1` makes the `uint8` and `int8` cases correct.
- PyTorch 2.5.1 on the same machine returns the correct result for this reproducer.

The failing generated C++ uses a partial vector load followed by an unmasked vector minimum and horizontal minimum, approximately:

```cpp
auto tmp0 = at::vec::Vectorized<uint8_t>::loadu(in_ptr0, 4);
tmp_acc0_vec = at::vec::minimum(tmp_acc0_vec, tmp0);
tmp_acc0 = at::vec::vec_reduce_all<uint8_t, 1>(
    [](auto& x, auto& y) { return at::vec::minimum(x, y); },
    tmp_acc0_vec);
```

A possible explanation is that inactive lanes from the partial vector load participate as zero in the horizontal minimum. This is only a code-generation hint; the reproducer and `simdlen=1` control are the confirmed observations.

### Environment

```text
PyTorch: 2.13.0
PyTorch git: cf30153c4c131c8164ee7798e5022d810682e2cb
Python: 3.12.10
OS: macOS 26.0 arm64
Machine: Apple M1 Pro / arm64
Backend: Inductor CPU
```

**Comment by lackjava-creator** [comment]
sure

Madhan Alagarsamy ***@***.***> 于 2026年7月30日周四 01:40写道：

**Comment by yeti107** [comment]
Hi! Is this still available or has it been fixed?

**Comment by cuiliaomei-beep** [comment]
This is not specific to Apple Silicon. On x86-64 (AVX2, Windows 11, MSVC; torch 2.14.0+cpu and nightly 2.15.0.dev20260911+cpu) the same defect appears once the reduction has 16 or more elements, and it affects more than `amin`:

```python
import torch

pos = torch.tensor([5, 1, 100, 3, 7, 8, 9, 13, 100, 120, 127, 2, 6, 4, 11, 12, 90, 80])
for dtype, x in ((torch.uint8, pos), (torch.int8, pos), (torch.int8, -pos)):
    x = x.to(dtype)
    for name, fn in (("min", torch.min), ("max", torch.max), ("argmin", torch.argmin), ("argmax", torch.argmax)):
        torch._dynamo.reset()
        e, c = fn(x).item(), torch.compile(fn)(x.clone()).item()
        if e != c:
            print(f"{str(dtype):12s} first={x[0].item():3d}  {name:6s} eager={e:4d}  inductor={c:4d}")
```

```
torch.uint8  first=  5  min    eager=   1  inductor=   0
torch.uint8  first=  5  argmin eager=   1  inductor=  -1
torch.int8   first=  5  min    eager=   1  inductor=   0
torch.int8   first=  5  argmin eager=   1  inductor=  -1
torch.int8   first= -5  max    eager=  -1  inductor=   0
torch.int8   first= -5  argmax eager=   1  inductor=  -1
```

New relative to the original report:

- **x86-64 is affected**; with 15 elements the results are correct, with 16 or more they are wrong (8-bit vectors have 32 lanes here, the reduction loads them 16 at a time).
- **`max` / `amax` of an all-negative int8 tensor returns 0**, the mirror image of `amin` on positive data. The value `0` does not occur in the input.
- **`argmin` / `argmax` return `-1`**, which is not a valid index; `torch.min(x, dim).indices` does the same. Code that indexes with the result reads the last element instead of failing.
- int16, int32, int64 and floating dtypes are correct on the same data, and so is CUDA (Triton).
- **The wrong value depends on the CPU.** The same script on two Linux x86-64 machines (GCC, torch 2.14.0 and nightly 2.15.0.dev20260920) gives `argmin = 0`, `max = 1` (all-negative int8, eager `-1`), `argmax = 0` on one, and `min = 0`, `argmin = -8473051087277178060`, `max = 0`, `argmax = 0` on the other; the last index is uninitialised memory, so the lanes that were not loaded are not even guaranteed to be zero.

The generated kernel shows the cause described above, a partial load followed by an unmasked combine:

```cpp
auto tmp0 = at::vec::Vectorized<uint8_t>::loadu(in_ptr0 + x0, static_cast<int64_t>(16));
tmp_acc0_vec = argmin_combine_vec<uint8_t, 1, 2, true>(tmp_acc0_vec, tmp0, x0);
```

`loadu(ptr, 16)` fills the other 16 lanes with zeros, and those lanes take part in the reduction (value 0, index -1). Passing the tail size to the combine step for the main loop as well (as the scalar-tail branch already does with `static_cast<int64_t>(1LL)`), or loading the remaining lanes with the reduction's identity, fixes all four operators.
